Mage/src/main/java/mage/game/GameImpl.java

```diff
@@ -1801,7 +1795,6 @@ protected void resolve() {
         } finally {
             if (top != null) {
                 state.getStack().remove(top, this); // seems partly redundant because move card from stack to grave is already done and the stack removed
-                rememberLKI(top.getSourceId(), Zone.STACK, top);
```

Comment:
That’s smell bad - why you remove that line? Now LKI will ignore any stack spells/abilities. So getSpellOrLKI never return LKI data.

Reply:
Looks like you just removed whole spell’s LKI. I don’t trust positive tests. If it’s really true then whole getSpellOrLKI must be removed as unused. So LKI will store only cards/permanents.

Reply:
That `rememberLKI` call never actually worked correctly. It was putting abilities into the Stack LKI under their source's ID. `getSpellOrLKI` didn't used to work, as proven by the counterspell added in `testWastescapeBattlemage`. Before this, it would sometimes find the ability instead of the actual spell. If I understood it correctly, spell LKI is added inside of `CardImpl.removeFromZone`, same as almost all LKI.

I could leave that line in (storing the LKI under the correct ID) but as far as I could tell, it isn't actually necessary.

Reply:
On a side note the `logger.error` branch I removed from `getSpellOrLKIStack` actually should probably not have been removed, that was a side effect of my investigations into the source of the problem. Fixing the bad LKI data fixed the actual problem.

Reply:
ok, I see. CardImpl.removeFromZone save spells only in LKI and never called for triggered abilities. Triggered abilities saved here (in old code). Now no more saved triggers in LKI.

If there are not any code that look at triggered ability in LKI then all fine.

If you look at 2014 commit then you can find such potentially broken code (looking at non spell ability in LKI):
![IMG_0631](https://github.com/user-attachments/assets/f199750a-d644-4316-9452-54506624d394).

Sacred Ground
> Whenever a spell or ability an opponent controls causes a land to be put into your graveyard from the battlefield, return that card to the battlefield.

I don’t see same usages in current master. But it can have another forms.

SO:
1. It’s fine to never save triggered ability to LKI (current PR);
2. Must return non spell assert to catch wrong data in LKI. But I recommend to put it in rememberLKI method to catch bad use cases immediately (example: if object instance of ability but non spell then raise wrong usage error).